File: src/main/java/minic/ui/text/MiniCTextStyles.java
package minic.ui.text;

import javafx.scene.Node;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.LinkedHashSet;
import java.util.List;

/**
 * Entry point for shared text style resolution.
 */
public final class MiniCTextStyles {
    private static final MiniCTextStyleResolver DEFAULT_RESOLVER = new DefaultResolver();

    private MiniCTextStyles() {}

    public static MiniCTextStyleResolver defaultResolver() {
        return DEFAULT_RESOLVER;
    }

    public static Collection<String> classes(MiniCTextStyleRole role, MiniCTextStyleState... states) {
        return DEFAULT_RESOLVER.styleClasses(role, Arrays.asList(states));
    }

    public static Collection<String> stateClasses(MiniCTextStyleState... states) {
        LinkedHashSet<String> classes = new LinkedHashSet<>();
        for (MiniCTextStyleState state : states) {
            classes.add(state.cssClass());
            classes.addAll(state.legacyClasses());
        }
        return List.copyOf(classes);
    }

    public static void addStateClasses(Collection<String> target, MiniCTextStyleState... states) {
        target.addAll(stateClasses(states));
    }

    public static void apply(Node node, MiniCTextStyleRole role, MiniCTextStyleState... states) {
        node.getStyleClass().addAll(classes(role, states));
    }

    private static final class DefaultResolver implements MiniCTextStyleResolver {
        @Override
        public Collection<String> styleClasses(MiniCTextStyleRole role, Collection<MiniCTextStyleState> states) {
            LinkedHashSet<String> classes = new LinkedHashSet<>();
            classes.add(role.cssClass());
            classes.addAll(role.legacyClasses());
            for (MiniCTextStyleState state : states == null ? List.<MiniCTextStyleState>of() : states) {
                classes.add(state.cssClass());
                classes.addAll(state.legacyClasses());
            }
            return new ArrayList<>(classes);
        }
    }
}

File: uiweb/src/text/MiniCTextStyles.ts
import type { JavaMirrorFile } from "../translation/javaMirror";
import type { MiniCTextStyleRole } from "./MiniCTextStyleRole";
import type { MiniCTextStyleState } from "./MiniCTextStyleState";
import { MiniCDefaultTextStyleResolver } from "./MiniCTextStyleResolver";
import type { MiniCResolvedTextStyle, MiniCTextStyleResolver } from "./MiniCTextStyleResolver";

export const miniCTextStylesMirror = {
  javaPath: "src/main/java/minic/uilocal/text/MiniCTextStyles.java",
  webPath: "uiweb/src/text/MiniCTextStyles.ts",
  packageName: "minic.uilocal.text",
  exportName: "MiniCTextStyles",
  kind: "class",
  imports: [
    "java.util.ArrayList",
    "java.util.Arrays",
    "java.util.Collection",
    "java.util.LinkedHashSet",
    "java.util.List",
    "javafx.scene.Node"
  ],
  fields: [
    {
      "name": "DEFAULT_RESOLVER",
      "signature": "private static final MiniCTextStyleResolver DEFAULT_RESOLVER="
    }
  ],
  methods: [
    {
      "name": "addStateClasses",
      "signature": "addStateClasses(Collection<String>target,MiniCTextStyleState... states)"
    },
    {
      "name": "apply",
      "signature": "apply(Node node,MiniCTextStyleRole role,MiniCTextStyleState... states)"
    },
    {
      "name": "classes",
      "signature": "classes(MiniCTextStyleRole role,MiniCTextStyleState... states)"
    },
    {
      "name": "defaultResolver",
      "signature": "defaultResolver()"
    },
    {
      "name": "stateClasses",
      "signature": "stateClasses(MiniCTextStyleState... states)"
    },
    {
      "name": "styleClasses",
      "signature": "styleClasses(MiniCTextStyleRole role,Collection<MiniCTextStyleState>states)"
    }
  ],
} as const satisfies JavaMirrorFile;

const DEFAULT_RESOLVER = new MiniCDefaultTextStyleResolver();

export class MiniCTextStyles {
  static readonly mirror = miniCTextStylesMirror;

  static defaultResolver(): MiniCTextStyleResolver {
    return DEFAULT_RESOLVER;
  }

  static classes(role: MiniCTextStyleRole, ...states: readonly MiniCTextStyleState[]): readonly string[] {
    return DEFAULT_RESOLVER.styleClasses(role, states);
  }

  static className(role: MiniCTextStyleRole, ...states: readonly MiniCTextStyleState[]): string {
    return this.classes(role, ...states).join(" ");
  }

  static stateClasses(...states: readonly MiniCTextStyleState[]): readonly string[] {
    return Array.from(new Set(states.flatMap((state) => [state.cssClass, ...state.legacyClasses])));
  }

  static addStateClasses(target: string[], ...states: readonly MiniCTextStyleState[]): void {
    target.push(...this.stateClasses(...states));
  }

  static resolve(role: MiniCTextStyleRole, ...states: readonly MiniCTextStyleState[]): MiniCResolvedTextStyle {
    return DEFAULT_RESOLVER.resolve(role, states);
  }
}

export default MiniCTextStyles;
